File: services/ai-agents/model_router.py
"""
AEGIS NEXUS — Python Model Router

High-performance routing layer for AI model inference. 
Attempts to use Anthropic Claude (with prompt caching) by default, 
falling back to a local Ollama instance if the API key is missing 
or a request failure occurs.
"""

from __future__ import annotations

import asyncio
import json
import logging
import os
import time
from dataclasses import dataclass, field
from typing import AsyncIterator, Literal

import anthropic
import httpx

logger = logging.getLogger("aegis.model_router")

# --- Configuration ---

ANTHROPIC_KEY   = os.getenv("ANTHROPIC_API_KEY", "")
ANTHROPIC_MODEL = os.getenv("ANTHROPIC_MODEL", "claude-opus-4-5")

OLLAMA_BASE     = os.getenv("OLLAMA_BASE_URL", "http://localhost:11434")
OLLAMA_MODEL    = os.getenv("OLLAMA_MODEL", "llama3.2")

Provider = Literal["anthropic", "ollama"]


@dataclass
class ModelResponse:
    """Unified response format for model completions."""
    text:          str
    model:         str
    provider:      Provider
    input_tokens:  int  = 0
    output_tokens: int  = 0
    cached:        bool = False
    duration_ms:   int  = 0


@dataclass
class StreamChunk:
    """Represents a single chunk in a streaming response."""
    text:     str
    done:     bool = False
    provider: Provider = "anthropic"


# --- Anthropic Provider ---

_anthropic_client: anthropic.AsyncAnthropic | None = None

def _get_anthropic() -> anthropic.AsyncAnthropic:
    """Returns a singleton instance of the Anthropic async client."""
    global _anthropic_client
    if _anthropic_client is None:
        _anthropic_client = anthropic.AsyncAnthropic(api_key=ANTHROPIC_KEY)
    return _anthropic_client


async def _anthropic_complete(
    system: str,
    messages: list[dict],
    max_tokens: int = 2048,
) -> ModelResponse:
    """Executes a completion request against the Anthropic API."""
    client = _get_anthropic()
    t0 = time.monotonic()

    response = await client.messages.create(
        model=ANTHROPIC_MODEL,
        max_tokens=max_tokens,
        system=[
            {
                "type": "text",
                "text": system,
                # Cache the system prompt to optimize costs for repeated agent calls
                "cache_control": {"type": "ephemeral"},
            }
        ],
        messages=messages,
    )

    text  = response.content[0].text if response.content else ""
    usage = response.usage
    cached = getattr(usage, "cache_read_input_tokens", 0) > 0

    return ModelResponse(
        text=text,
        model=response.model,
        provider="anthropic",
        input_tokens=usage.input_tokens,
        output_tokens=usage.output_tokens,
        cached=cached,
        duration_ms=int((time.monotonic() - t0) * 1000),
    )


async def _anthropic_stream(
    system: str,
    messages: list[dict],
    max_tokens: int = 2048,
) -> AsyncIterator[StreamChunk]:
    """Streams a completion response from the Anthropic API."""
    client = _get_anthropic()

    async with client.messages.stream(
        model=ANTHROPIC_MODEL,
        max_tokens=max_tokens,
        system=[
            {"type": "text", "text": system, "cache_control": {"type": "ephemeral"}}
        ],
        messages=messages,
    ) as stream:
        async for text in stream.text_stream:
            yield StreamChunk(text=text, provider="anthropic")

    yield StreamChunk(text="", done=True, provider="anthropic")


# --- Ollama Provider ---

async def _ollama_complete(
    system: str,
    messages: list[dict],
    max_tokens: int = 2048,
) -> ModelResponse:
    """Executes a completion request against a local Ollama instance."""
    t0 = time.monotonic()

    payload = {
        "model":    OLLAMA_MODEL,
        "messages": [{"role": "system", "content": system}, *messages],
        "stream":   False,
        "options":  {"num_predict": max_tokens},
    }

    async with httpx.AsyncClient(timeout=120.0) as client:
        resp = await client.post(f"{OLLAMA_BASE}/api/chat", json=payload)
        resp.raise_for_status()
        data = resp.json()

    return ModelResponse(
        text=data["message"]["content"],
        model=data.get("model", OLLAMA_MODEL),
        provider="ollama",
        input_tokens=data.get("prompt_eval_count", 0),
        output_tokens=data.get("eval_count", 0),
        cached=False,
        duration_ms=int((time.monotonic() - t0) * 1000),
    )


async def _ollama_stream(
    system: str,
    messages: list[dict],
    max_tokens: int = 2048,
) -> AsyncIterator[StreamChunk]:
    """Streams a completion response from a local Ollama instance."""
    payload = {
        "model":    OLLAMA_MODEL,
        "messages": [{"role": "system", "content": system}, *messages],
        "stream":   True,
        "options":  {"num_predict": max_tokens},
    }

    async with httpx.AsyncClient(timeout=120.0) as client:
        async with client.stream("POST", f"{OLLAMA_BASE}/api/chat", json=payload) as resp:
            resp.raise_for_status()
            async for line in resp.aiter_lines():
                if not line.strip():
                    continue
                try:
                    obj = json.loads(line)
                    content = obj.get("message", {}).get("content", "")
                    if content:
                        yield StreamChunk(text=content, provider="ollama")
                    if obj.get("done"):
                        break
                except json.JSONDecodeError:
                    pass

    yield StreamChunk(text="", done=True, provider="ollama")


async def ollama_available() -> bool:
    """Performs a health check to determine if the Ollama service is reachable."""
    try:
        async with httpx.AsyncClient(timeout=2.0) as client:
            resp = await client.get(f"{OLLAMA_BASE}/api/tags")
            return resp.status_code == 200
    except Exception:
        return False


# --- Core Router ---

async def complete(
    system: str,
    messages: list[dict],
    max_tokens: int = 2048,
    prefer: Provider | None = None,
) -> ModelResponse:
    """
    Unified entry point for model completions.
    
    Priority Logic:
    1. Anthropic (if API key is present and 'prefer' is not 'ollama')
    2. Local Ollama (fallback)
    """
    use_anthropic = ANTHROPIC_KEY and prefer != "ollama"

    if use_anthropic:
        try:
            result = await _anthropic_complete(system, messages, max_tokens)
            logger.info(
                "anthropic/%s in=%d out=%d cached=%s dur=%dms",
                result.model, result.input_tokens, result.output_tokens,
                result.cached, result.duration_ms,
            )
            return result
        except anthropic.APIError as e:
            logger.warning("Anthropic failed (%s) — falling back to Ollama", e)

    logger.info("ollama/%s", OLLAMA_MODEL)
    return await _ollama_complete(system, messages, max_tokens)


async def stream(
    system: str,
    messages: list[dict],
    max_tokens: int = 2048,
    prefer: Provider | None = None,
) -> AsyncIterator[StreamChunk]:
    """
    Unified entry point for streaming model completions.
    
    Follows the same priority logic as the `complete` function.
    Yields `StreamChunk` objects until the final chunk (marked as `done=True`).
    """
    use_anthropic = ANTHROPIC_KEY and prefer != "ollama"

    if use_anthropic:
        try:
            async for chunk in _anthropic_stream(system, messages, max_tokens):
                yield chunk
            return
        except anthropic.APIError as e:
            logger.warning("Anthropic stream failed (%s) — falling back to Ollama", e)

    async for chunk in _ollama_stream(system, messages, max_tokens):
        yield chunk


def active_provider() -> dict:
    """Returns the current configuration and active status of supported providers."""
    return {
        "anthropic": {
            "configured": bool(ANTHROPIC_KEY),
            "model":      ANTHROPIC_MODEL,
        },
        "ollama": {
            "base":  OLLAMA_BASE,
            "model": OLLAMA_MODEL,
        },
        "active": "anthropic" if ANTHROPIC_KEY else "ollama",
    }


async def _smoke_test():
    """Simple internal validation to verify the router is functional."""
    logging.basicConfig(level=logging.INFO)
    result = await complete(
        system="You are a test assistant. Reply with exactly: AEGIS MODEL ROUTER OK",
        messages=[{"role": "user", "content": "ping"}],
        max_tokens=20,
    )
    print(f"Provider: {result.provider} | Model: {result.model}")
    print(f"Response: {result.text.strip()}")
    print(f"Tokens in/out: {result.input_tokens}/{result.output_tokens} | Cached: {result.cached}")


if __name__ == "__main__":
    asyncio.run(_smoke_test())

File: apps/web/lib/model-router.ts
/**
 * AEGIS Model Router
 * 
 * Orchestrates AI model requests by trying Anthropic (Claude) first 
 * and falling back to a local Ollama instance if the API key is missing 
 * or the request fails. Designed for server-side execution within 
 * Next.js API routes.
 */

import Anthropic from "@anthropic-ai/sdk";

// --- Types ---

/** Represents a single message in a model conversation. */
export interface ModelMessage {
  role: "user" | "assistant";
  content: string;
}

/** Configuration for a model inference request. */
export interface ModelRequest {
  /** The system prompt to guide agent behavior. */
  system: string;
  /** Conversation history. */
  messages: ModelMessage[];
  /** Maximum number of tokens to generate. */
  maxTokens?: number;
  /** Whether to stream the response. */
  stream?: boolean;
}

/** Unified response format across different providers. */
export interface ModelResponse {
  /** The generated text content. */
  text: string;
  /** The specific model name used. */
  model: string;
  /** The provider that handled the request. */
  provider: "anthropic" | "ollama";
  /** Number of tokens in the input prompt. */
  inputTokens: number;
  /** Number of tokens in the output response. */
  outputTokens: number;
  /** Whether the request hit a provider-side cache. */
  cached: boolean;
}

// --- Configuration ---

const ANTHROPIC_KEY   = process.env.ANTHROPIC_API_KEY ?? "";
const OLLAMA_BASE     = process.env.OLLAMA_BASE_URL ?? "http://localhost:11434";
const OLLAMA_MODEL    = process.env.OLLAMA_MODEL ?? "llama3.2";
const DEFAULT_MODEL   = process.env.ANTHROPIC_MODEL ?? "claude-opus-4-5";

const anthropicClient = ANTHROPIC_KEY ? new Anthropic({ apiKey: ANTHROPIC_KEY }) : null;

// --- Anthropic Provider ---

/**
 * Internal helper to call the Anthropic Messages API.
 */
async function callAnthropic(req: ModelRequest): Promise<ModelResponse> {
  if (!anthropicClient) throw new Error("ANTHROPIC_API_KEY not set");

  const response = await anthropicClient.messages.create({
    model: DEFAULT_MODEL,
    max_tokens: req.maxTokens ?? 2048,
    system: [
      {
        type: "text",
        text: req.system,
        // Cache the system prompt across requests for the same agent role
        cache_control: { type: "ephemeral" },
      },
    ],
    messages: req.messages,
  });

  const block = response.content[0];
  const text  = block.type === "text" ? block.text : "";
  const usage = response.usage as Anthropic.Usage & { cache_read_input_tokens?: number };

  return {
    text,
    model: response.model,
    provider: "anthropic",
    inputTokens: usage.input_tokens,
    outputTokens: usage.output_tokens,
    cached: (usage.cache_read_input_tokens ?? 0) > 0,
  };
}

/**
 * Streams a response from Anthropic.
 */
export async function streamAnthropic(
  req: ModelRequest,
  onChunk: (text: string) => void,
  onDone: (response: Omit<ModelResponse, "text">) => void,
): Promise<void> {
  if (!anthropicClient) throw new Error("ANTHROPIC_API_KEY not set");

  const stream = anthropicClient.messages.stream({
    model: DEFAULT_MODEL,
    max_tokens: req.maxTokens ?? 2048,
    system: [
      {
        type: "text",
        text: req.system,
        cache_control: { type: "ephemeral" },
      },
    ],
    messages: req.messages,
  });

  for await (const event of stream) {
    if (
      event.type === "content_block_delta" &&
      event.delta.type === "text_delta"
    ) {
      onChunk(event.delta.text);
    }
  }

  const final = await stream.finalMessage();
  const usage = final.usage as Anthropic.Usage & { cache_read_input_tokens?: number };
  onDone({
    model: final.model,
    provider: "anthropic",
    inputTokens: usage.input_tokens,
    outputTokens: usage.output_tokens,
    cached: (usage.cache_read_input_tokens ?? 0) > 0,
  });
}

// --- Ollama Provider ---

/**
 * Internal helper to call a local Ollama instance via its REST API.
 */
async function callOllama(req: ModelRequest): Promise<ModelResponse> {
  const payload = {
    model: OLLAMA_MODEL,
    messages: [
      { role: "system", content: req.system },
      ...req.messages,
    ],
    stream: false,
    options: { num_predict: req.maxTokens ?? 2048 },
  };

  const res = await fetch(`${OLLAMA_BASE}/api/chat`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
    signal: AbortSignal.timeout(120_000),
  });

  if (!res.ok) {
    const body = await res.text();
    throw new Error(`Ollama error ${res.status}: ${body}`);
  }

  const data = (await res.json()) as {
    message: { content: string };
    model: string;
    prompt_eval_count?: number;
    eval_count?: number;
  };

  return {
    text: data.message.content,
    model: data.model,
    provider: "ollama",
    inputTokens: data.prompt_eval_count ?? 0,
    outputTokens: data.eval_count ?? 0,
    cached: false,
  };
}

/**
 * Streams a response from a local Ollama instance.
 */
export async function streamOllama(
  req: ModelRequest,
  onChunk: (text: string) => void,
  onDone: (response: Omit<ModelResponse, "text">) => void,
): Promise<void> {
  const payload = {
    model: OLLAMA_MODEL,
    messages: [
      { role: "system", content: req.system },
      ...req.messages,
    ],
    stream: true,
  };

  const res = await fetch(`${OLLAMA_BASE}/api/chat`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
    signal: AbortSignal.timeout(120_000),
  });

  if (!res.ok || !res.body) throw new Error(`Ollama stream failed: ${res.status}`);

  const reader = res.body.getReader();
  const decoder = new TextDecoder();
  let inputTokens = 0;
  let outputTokens = 0;
  let model = OLLAMA_MODEL;

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;
    const lines = decoder.decode(value).split("\n").filter(Boolean);
    for (const line of lines) {
      try {
        const obj = JSON.parse(line) as {
          message?: { content: string };
          model?: string;
          done?: boolean;
          prompt_eval_count?: number;
          eval_count?: number;
        };
        if (obj.message?.content) onChunk(obj.message.content);
        if (obj.model) model = obj.model;
        if (obj.done) {
          inputTokens  = obj.prompt_eval_count ?? 0;
          outputTokens = obj.eval_count ?? 0;
        }
      } catch { /* partial JSON */ }
    }
  }

  onDone({ model, provider: "ollama", inputTokens, outputTokens, cached: false });
}

// --- Router Functions ---

/**
 * Smart router: attempts to fulfill the request via Anthropic, 
 * falling back to Ollama on failure or if the key is missing.
 */
export async function routedComplete(req: ModelRequest): Promise<ModelResponse> {
  if (ANTHROPIC_KEY) {
    try {
      const result = await callAnthropic(req);
      console.log(
        `[router] anthropic/${result.model} — in:${result.inputTokens} out:${result.outputTokens} cached:${result.cached}`,
      );
      return result;
    } catch (err) {
      console.warn("[router] Anthropic failed, falling back to Ollama:", err);
    }
  }

  console.log(`[router] ollama/${OLLAMA_MODEL} (Anthropic key ${ANTHROPIC_KEY ? "failed" : "not set"})`);
  return callOllama(req);
}

/**
 * Streaming router: initiates a stream from the best available provider.
 * Returns the name of the active provider.
 */
export async function routedStream(
  req: ModelRequest,
  onChunk: (text: string) => void,
  onDone: (response: Omit<ModelResponse, "text">) => void,
): Promise<"anthropic" | "ollama"> {
  if (ANTHROPIC_KEY) {
    try {
      await streamAnthropic(req, onChunk, onDone);
      return "anthropic";
    } catch (err) {
      console.warn("[router] Anthropic stream failed, falling back to Ollama:", err);
    }
  }
  await streamOllama(req, onChunk, onDone);
  return "ollama";
}

/**
 * Returns the current availability status of supported model providers.
 */
export function availableProviders(): { anthropic: boolean; ollama: string } {
  return {
    anthropic: Boolean(ANTHROPIC_KEY),
    ollama: `${OLLAMA_BASE} (model: ${OLLAMA_MODEL})`,
  };
}
